src/capture/image/png_writer.cpp

```diff
@@ -0,0 +1,260 @@
+
+	// Do not change the below settings; they are parameters for the zlib
+	// compression library and changing them might result in invalid PNG
+	// files.
+	constexpr auto default_mem_level = 8;
+	png_set_compression_mem_level(png_ptr, default_mem_level);
+
+	constexpr auto default_window_bits = 15;
+	png_set_compression_window_bits(png_ptr, default_window_bits);
+
+	png_set_compression_strategy(png_ptr, Z_DEFAULT_STRATEGY);
+	png_set_compression_method(png_ptr, Z_DEFLATED);
+}
+
+void PngWriter::WritePngInfo(const ImageInfo& image_info,
+                             const std::optional<ImageInfo>& source_image_info,
+                             const bool is_paletted, const uint8_t* palette_data)
+{
+	assert(png_ptr);
+	assert(png_info_ptr);
+
+	constexpr auto png_bit_depth = 8;
+	const auto png_color_type    = is_paletted ? PNG_COLOR_TYPE_PALETTE
+	                                           : PNG_COLOR_TYPE_RGB;
+	png_set_IHDR(png_ptr,
+	             png_info_ptr,
+	             image_info.width,
+	             image_info.height,
+	             png_bit_depth,
+	             png_color_type,
+	             PNG_INTERLACE_NONE,
+	             PNG_COMPRESSION_TYPE_DEFAULT,
+	             PNG_FILTER_TYPE_DEFAULT);
+
+	if (is_paletted) {
+		constexpr auto NumPaletteEntries     = 256;
+		png_color palette[NumPaletteEntries] = {};
+
+		for (auto i = 0; i < NumPaletteEntries; ++i) {
+			palette[i].red   = palette_data[i * 4 + 0];
+			palette[i].green = palette_data[i * 4 + 1];
+			palette[i].blue  = palette_data[i * 4 + 2];
+		}
+		png_set_PLTE(png_ptr, png_info_ptr, palette, NumPaletteEntries);
+	}
+
+#ifdef PNG_gAMA_SUPPORTED
+	// It's not strictly necessary to write this chunk, but it's recommended
+	// by the spec.
+	png_set_gAMA(png_ptr, png_info_ptr, 1 / 2.2);
+#endif
+
+#ifdef PNG_pHYs_SUPPORTED
+	// "The pHYs chunk specifies the intended pixel size or aspect ratio for
+	// display of the image."
+	//
+	// "If the pHYs chunk is not present, pixels are assumed to be square,
+	// and the physical size of each pixel is unspecified."
+	//
+	// Although as of now pretty much all programs ignore the pHYs chunk and
+	// simply assume square pixels, we're writing the correct pixel aspect
+	// ratio in the hope that in the future applications will handle the
+	// pHYs chunk appropriately.
+	//
+	// Source:
+	//   Portable Network Graphics (PNG) Specification (Second Edition)
+	//   https://www.w3.org/TR/2003/REC-PNG-20031110/#11pHYs)
+	//
+	const uint32_t pixels_per_unit_x = static_cast<uint32_t>(
+	        image_info.pixel_aspect_ratio.Num());
+
+	const uint32_t pixels_per_unit_y = static_cast<uint32_t>(
+	        image_info.pixel_aspect_ratio.Denom());
+
+	// "When the unit specifier is 0, the pHYs chunk defines pixel aspect
+	// ratio only; the actual size of the pixels remains unspecified."
+	auto unit_type = 0;
+
+	png_set_pHYs(png_ptr, png_info_ptr, pixels_per_unit_x, pixels_per_unit_y, unit_type);
```

Comment:
:ok_hand: very nice that this is included!

Reply:
Maybe we'll live to see non-square pixel aspect ratio support in common image viewers! Hope dies last! 😎